File: src/ou2.c
// -*- C++ -*-

#include <R.h>
#include <Rmath.h>
#include <Rdefines.h>
#include "internal.h"

// simple 2D Ornstein-Uhlenbeck process simulation
static void sim_ou2
(
 double *x1, double *x2,
 double alpha1, double alpha2, double alpha3, double alpha4,
 double sigma1, double sigma2, double sigma3
 ) {
  double eps[2], xnew[2];

  eps[0] = rnorm(0,1);
  eps[1] = rnorm(0,1);

  xnew[0] = alpha1*(*x1)+alpha3*(*x2)+sigma1*eps[0];
  xnew[1] = alpha2*(*x1)+alpha4*(*x2)+sigma2*eps[0]+sigma3*eps[1];

  *x1 = xnew[0];
  *x2 = xnew[1];
}

// simple 2D Ornstein-Uhlenbeck process transition density
// transition (x1,x2) -> (z1,z2) in 1 unit of time
static double dens_ou2
(
 double x1, double x2, double z1, double z2,
 double alpha1, double alpha2, double alpha3, double alpha4,
 double sigma1, double sigma2, double sigma3, int give_log
 ) {
  double eps[2], val;

  // compute residuals
  eps[0] = z1-alpha1*x1-alpha3*x2;
  eps[1] = z2-alpha2*x1-alpha4*x2;

  // backsolve
  eps[0] /= sigma1;
  eps[1] -= sigma2*eps[0];
  eps[1] /= sigma3;

  val = dnorm(eps[0],0.0,1.0,1)+dnorm(eps[1],0.0,1.0,1)-log(sigma1)-log(sigma3);
  return ((give_log) ? val : exp(val));
}

#define ALPHA1     (p[parindex[0]])
#define ALPHA2     (p[parindex[1]])
#define ALPHA3     (p[parindex[2]])
#define ALPHA4     (p[parindex[3]])
#define SIGMA1     (p[parindex[4]])
#define SIGMA2     (p[parindex[5]])
#define SIGMA3     (p[parindex[6]])
#define TAU        (p[parindex[7]])

#define X1    (stateindex[0])
#define X2    (stateindex[1])

#define Y1    (y[obsindex[0]])
#define Y2    (y[obsindex[1]])

#define V11   (f[vmatindex[0]])
#define V21   (f[vmatindex[1]])
#define V12   (f[vmatindex[2]])
#define V22   (f[vmatindex[3]])

// onestep simulator for use in 'discrete.time.sim' plug-in
void _ou2_step
(
 double *x, const double *p,
 const int *stateindex, const int *parindex, const int *covindex,
 const double *covars, double t, double deltat
 ) {
  sim_ou2(&x[X1],&x[X2],ALPHA1,ALPHA2,ALPHA3,ALPHA4,SIGMA1,SIGMA2,SIGMA3);
}

// onestep transition probability density for use in 'onestep.dens' plug-in
// transition from x to z as time goes from t1 to t2
void _ou2_pdf
(
 double *f,
 double *x, double *z, double t1, double t2, const double *p,
 const int *stateindex, const int *parindex, const int *covindex,
 const double *covars
 ) {
  if (t2-t1 != 1)
    err("ou2_pdf error: transitions must be consecutive");
  f[0] = dens_ou2(x[X1],x[X2],z[X1],z[X2],ALPHA1,ALPHA2,ALPHA3,ALPHA4,
                  SIGMA1,SIGMA2,SIGMA3,1);
}

void _ou2_skel
(
 double *f, double *x, double *p,
 int *stateindex, int *parindex, int *covindex,
 double *covars, double t
 ) {
  f[X1] = ALPHA1*x[X1]+ALPHA3*x[X2];
  f[X2] = ALPHA2*x[X1]+ALPHA4*x[X2];
}

// bivariate normal measurement error density
void _ou2_dmeasure
(
 double *lik, double *y, double *x, double *p, int give_log,
 int *obsindex, int *stateindex, int *parindex, int *covindex,
 double *covar, double t
 ) {
  double sd = fabs(TAU);
  double f = 0.0;
  f += (ISNA(Y1)) ? 0.0 : dnorm(Y1,x[X1],sd,1);
  f += (ISNA(Y2)) ? 0.0 : dnorm(Y2,x[X2],sd,1);
  *lik = (give_log) ? f : exp(f);
}

// bivariate normal measurement error simulator
void _ou2_rmeasure
(
 double *y, double *x, double *p,
 int *obsindex, int *stateindex, int *parindex, int *covindex,
 double *covar, double t
 ) {
  double sd = fabs(TAU);
  Y1 = rnorm(x[X1],sd);
  Y2 = rnorm(x[X2],sd);
}

// bivariate normal measurement error expectation
void _ou2_emeasure
(
 double *y, double *x, double *p,
 int *obsindex, int *stateindex, int *parindex, int *covindex,
 double *covar, double t
 ) {
  Y1 = x[X1];
  Y2 = x[X2];
}

// bivariate normal measurement variance
void _ou2_vmeasure
(
 double *f, double *x, double *p,
 int *vmatindex, int *stateindex, int *parindex, int *covindex,
 double *covar, double t
 ) {
  double sd = fabs(TAU);
  V11 = V22 = sd*sd;
  V12 = V21 = 0;
}

#undef ALPHA1
#undef ALPHA2
#undef ALPHA3
#undef ALPHA4
#undef SIGMA1
#undef SIGMA2
#undef SIGMA3
#undef TAU

#undef X1
#undef X2
#undef Y1
#undef Y2

#undef V11
#undef V21
#undef V12
#undef V22

File: R/ou2.R
##' Two-dimensional discrete-time Ornstein-Uhlenbeck process
##'
##' \code{ou2()} constructs a \sQuote{pomp} object encoding a bivariate discrete-time Ornstein-Uhlenbeck process with noisy observations.
##'
##' If the state process is \eqn{X(t) = (x_{1}(t),x_{2}(t))}, then
##' \deqn{X(t+1) = \alpha X(t) + \sigma \epsilon(t),}
##' where \eqn{\alpha} and \eqn{\sigma} are 2x2 matrices,
##' \eqn{\sigma} is lower-triangular, and
##' \eqn{\epsilon(t)} is standard bivariate normal.
##' The observation process is \eqn{Y(t) = (y_1(t),y_2(t))}, where
##' \eqn{y_i(t) \sim \mathrm{normal}(x_i(t),\tau)}.
##'
##' @name ou2
##' @rdname ou2
##' @include simulate.R
##' @docType data
##' @keywords models pomp_datasets
##' @family pomp examples
##' @return A \sQuote{pomp} object with simulated data.
##' @example examples/ou2.R
##'
NULL

##' @rdname ou2
##' @param alpha_1,alpha_2,alpha_3,alpha_4 entries of the \eqn{alpha} matrix, in column-major order.
##' That is, \code{alpha_2} is in the lower-left position.
##' @param sigma_1,sigma_2,sigma_3 entries of the lower-triangular \eqn{sigma} matrix.
##' \code{sigma_2} is the entry in the lower-left position.
##' @param tau measurement error s.d.
##' @param x1_0,x2_0 latent variable values at time \code{t0}
##' @param t0 the zero time
##' @param times vector of observation times
##'
##' @export
ou2 <- function (
  alpha_1 = 0.8, alpha_2 = -0.5, alpha_3 = 0.3, alpha_4 = 0.9,
  sigma_1 = 3, sigma_2 = -0.5, sigma_3 = 2,
  tau = 1,
  x1_0 = -3, x2_0 = 4,
  times = 1:100, t0 = 0
)
{

  simulate(
    times=times, t0=t0,
    seed=787832394,
    rprocess=discrete_time("_ou2_step"),
    dprocess="_ou2_pdf",
    dmeasure = "_ou2_dmeasure",
    rmeasure = "_ou2_rmeasure",
    skeleton = map("_ou2_skel",delta.t=1),
    PACKAGE="pomp",
    paramnames = c(
      "alpha_1","alpha_2","alpha_3","alpha_4",
      "sigma_1","sigma_2","sigma_3",
      "tau"
    ),
    statenames = c("x1","x2"),
    obsnames = c("y1","y2"),
    params=c(
      alpha_1=alpha_1, alpha_2=alpha_2, alpha_3=alpha_3, alpha_4=alpha_4,
      sigma_1=sigma_1, sigma_2=sigma_2, sigma_3=sigma_3,
      tau=tau,
      x1_0=x1_0, x2_0=x2_0
    )
  )

}
